### PR TITLE
More tests & better checking of extraction results.

### PR DESCRIPTION
**Pull request by TrevorHansen**
In creating the nicer ILP extractor (#16) I've had lots of annoying rare errors that have been hard to pin down. So I've made a fuzzer that randomly generates egraphs, then I run different extractors on those random egraphs. 

Because the ILP extractors (ilp-cbc & faster-ilp-cbc) are optimal for DAG-extraction, the DAG-extractions from the other extractiors should never be cheaper on a given egraph. The effect of all this is that the extractors get tested on tens of millions of egraphs. It's found lots of problems with the extractors. 

That's the back story.

Anyway, this PR contains some egraph that the fuzzer has generated, some that I've handcrafted to check extreme cases and better testing of extraction results.

These extra tests reveal problems with the faster-greedy-dag & global-greedy-dag extractors. Both of which fail on some instances. 

The faster-greedy-dag returns extractions with cycles and the global-greedy-dag sometimes returns extractions with the wrong nodes selected (i.e. a node will be selected for the wrong class). I'll create cases for these.

### CODE REVIEW

**Review by oflatt — APPROVED** [review]
Looks like a great change to me!

File: src/main.rs [hunk 1]
```diff
@@ -78,11 +78,9 @@ fn main() {
 
     let start_time = std::time::Instant::now();
     let result = extractor.extract(&egraph, &egraph.root_eclasses);
+    result.check(&egraph);
```
Comment (oflatt) [inline src/main.rs:81]: should this check happen after we measure the time? (line 83)

**Comment by oflatt** [comment]
Does this PR contain the cases where `global-greedy-dag` fails, or not?

**Comment by TrevorHansen** [comment]
> Does this PR contain the cases where `global-greedy-dag` fails, or not?

Yes, the global-greedy-dag extractor "fails", in that it chooses nodes from the wrong class, on some of the egraphs already in there. We just notice it now because there's more checking.

**Comment by oflatt** [comment]
Oh, we should change the CI so that it actually runs all the extractors then

**Pull request merged**